File: java/Strings/RabinKarp.java
public class RabinKarp {
    public static int rabinKarp(String text, String pattern) {
        if (text == null || pattern == null) return -1;
        int n = text.length(), m = pattern.length();
        if (m > n) return -1;

        int base = 256;               
        int mod = 101;              

        int patternHash = 0;          
        int textHash = 0;             
        int h = 1;                    

        
        for (int i = 0; i < m - 1; i++)
            h = (h * base) % mod;

        
        for (int i = 0; i < m; i++) {
            patternHash = (base * patternHash + pattern.charAt(i)) % mod;
            textHash = (base * textHash + text.charAt(i)) % mod;
        }

        
        for (int i = 0; i <= n - m; i++) {
           
            if (patternHash == textHash) {
               
                int j;
                for (j = 0; j < m; j++) {
                    if (text.charAt(i + j) != pattern.charAt(j))
                        break;
                }
                if (j == m) return i; 
            }

           
            if (i < n - m) {
                textHash = (base * (textHash - text.charAt(i) * h) + text.charAt(i + m)) % mod;

                
                if (textHash < 0)
                    textHash += mod;
            }
        }
        return -1; 
    }

    public static void main(String[] args) {
        String text = "ABCCDDAEFG";
        String pattern = "CDD";
        int index = rabinKarp(text, pattern);
        System.out.println("Pattern found at index: " + index); 

        System.out.println("Pattern found at index: " + rabinKarp("hello", "ll")); 
        System.out.println("Pattern found at index: " + rabinKarp("aaaaa", "bba")); 
    }
}

File: rust/rabin_karp.rs
fn rabin_karp(text: &str, pattern: &str) {
    let d: u64 = 256;        // Number of characters in input alphabet
    let q: u64 = 101;        // A prime number for modulo
    let m = pattern.len();
    let n = text.len();

    if m > n {
        println!("Pattern is longer than text — no match.");
        return;
    }

    let text_bytes = text.as_bytes();
    let pattern_bytes = pattern.as_bytes();

    let mut p: u64 = 0; // hash for pattern
    let mut t: u64 = 0; // hash for text window
    let mut h: u64 = 1; // d^(m-1) % q

    // Precompute h = pow(d, m-1) % q
    for _ in 0..m - 1 {
        h = (h * d) % q;
    }

    // Compute initial hash values for pattern and first window
    for i in 0..m {
        p = (d * p + pattern_bytes[i] as u64) % q;
        t = (d * t + text_bytes[i] as u64) % q;
    }

    // Slide the pattern over the text
    for i in 0..=(n - m) {
        // Check if hashes match
        if p == t {
            // Double-check characters
            if &text[i..i + m] == pattern {
                println!("Pattern found at index {}", i);
            }
        }

        // Compute hash for next window
        if i < n - m {
            t = (d * (t + q - (text_bytes[i] as u64 * h) % q) + text_bytes[i + m] as u64) % q;
        }
    }
}

fn main() {
    let text = "ABCCDDAEFG";
    let pattern = "CDD";

    rabin_karp(text, pattern);
}
